lib/SPIRV/SPIRVReader.cpp

```diff
@@ -2475,6 +2475,36 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
         BV, Builder.CreateIntrinsic(Intrinsic::expect, RetTy, {Val, ExpVal}));
   }
 
+  case OpUntypedPrefetchKHR: {
+    // Do the same as transOCLBuiltinFromExtInst() but for OpUntypedPrefetchKHR.
+    auto *BC = static_cast<SPIRVUntypedPrefetchKHR *>(BV);
+
+    std::vector<Type *> ArgTypes =
+        transTypeVector(BC->getValueTypes(BC->getArguments()), true);
+    Type *RetTy = Type::getVoidTy(*Context);
+
+    std::string MangledName =
+        getSPIRVFriendlyIRFunctionName(OpenCLLIB::Prefetch, ArgTypes, RetTy);
```

Comment:
I started to remember (missed it in our chat). We need to map the instruction to llvm.prefetch intrinsic. So probably the logic should be:
1. if RW, locality and cache type optional parameters are set, then map the instruction to llvm.prefetch;
2. else if some of the optional parameters present, then we should emit SPIR-V friendly LLVM IR call (btw, @alan-baker I believe some routine lines should be added to the spec like: "if Cache Type present, then other optional parameters must be set");
3. if no optional parameters are set - map it to OpenCL prefetch call.

Note, No.2 is discussible, we may want to still emit llvm intrinsic but with some default values set. In the case No.2 RW will always be set, which is good as we won't need to guess a default value, locality can be defaulted to '0'. Not sure though for Cache Type, '1' seems reasonable to me. Would be nice to hear from @alan-baker @bashbaug and @svenvh about that.

Reply:
We could add that restriction in the future if that is the only practical use case. I left it vague because I am not familiar with generating the equivalent LLVM intrinsic.

As for defaults I simply tried to provide the same arguments as the LLVM intrinsics, though I can't imagine how instruction cache fetching could be done via OpenCL C. So for Cache Type I'd expect it's always safe to default to data.

Reply:
About p.1, I'm not sure we should use `llvm.prefetch` intrinsic for now because it does not contain the number of bytes to prefetch field.
The similar concern is for forward translation. Although `llvm.prefetch` contains `RW`, `Locality` and `Cache Type` fields, it does not have a number of bytes to prefetch, which is a mandatory field in `OpUntypedPrefetchKHR` instruction.

Reply:
> it does not contain the number of bytes to prefetch field

Indeed. Probably it can go to another PR, but we can add the following logic:
1. (as before) if no optional parameters are set - map it to OpenCL prefetch call;
2. if one of the optional parameters (RW, locality or cache type) optional parameters is set and number of bytes = 64, then map it on llvm.prefetch (on x86-64 prefetch would prefetch 64 bytes, other magic number is also possible);
3. if optional parameter present and number of bytes != 64 - leave it as SPIR-V friendly call.